### PR TITLE
[WIP] Data importer

### PR DESCRIPTION
**Pull request by SchrodingersGat**
This PR represents a major change in approach to how we import / export data.

Previously we have relied on external libraries for importing / exporting - which all come with various downsides.

The main idea of this PR is to utilize the existing DRF API framework for both importing and exporting of data. This should provide the following major advantages:

- Significantly reduce code duplication
- Hook in to existing API framework which is well tested 
- Improve speed of import / export
- Allow finer control over import / export processes
- Allow specification of custom import hooks 
- Expose import / export processes to custom plugins

### Framework

Uses existing DRF serializers, just need to register with the `@register_importer()` decorator. 

### Related Issues

- Closes https://github.com/inventree/InvenTree/pull/6792
- Closes https://github.com/inventree/InvenTree/issues/6376
- Closes https://github.com/inventree/InvenTree/issues/6389
- Closes https://github.com/inventree/InvenTree/issues/6955
- Closes https://github.com/inventree/InvenTree/issues/6956
- Closes https://github.com/inventree/InvenTree/issues/7431
- Closes: https://github.com/inventree/InvenTree/issues/6321
- Ref: https://github.com/inventree/InvenTree/issues/5571
- Ref: https://github.com/inventree/InvenTree/issues/6664
- Ref: https://github.com/inventree/InvenTree/issues/6647
- Ref: https://github.com/inventree/InvenTree/issues/3101

### TODO

- [x] Remove the `APIDownloadMixin` class (and all references)
- [x] Remove the `download_queryset` method in a many existing views
- [x] Use database model `verbose_name` if serializer does not provide a `label` attribute
- [x] Add / update unit testing for new dataset export approach
- [x] Add worker task to cleanup old import / export sessions
- [x] Ensure all 'required' fields are mapped before progressing
- [x] Add data import section to "admin center"
- [x] Add in-line rendering of related model fields
- [x] Improve query efficiency of "import session" endpoint (takes ~1 second!?)
- [x] Add unit testing for new dataset import approach

### Further Work

- [ ] Support "default values" when importing dataset
- [ ] Allow user to navigate back to previous data steps
- [ ] Add unit test for cleanup backup task
- [ ] Ensure user has correct permissions before creating a new import session
- [ ] Expose import / export functions to plugins
- [ ] Allow plugins to define custom export fields / processes
- [ ] Allow bulk data export action to django admin interface
- [ ] Replace existing "BOM Import" tool with new importer
- [ ] Replace existing "BOM export" tool with new exporter
- [ ] Replace existing "part import" tool with new importer
- [ ] Replace existing "purchase order import" tool with new importer
- [ ] Remove `django-import-export` framework entirely
- [ ] Allow bulk "update" via import (override current records, not create new ones)
- [ ] Remove all custom import UI template code (html templates, legacy)
- [ ] Check that all existing "django-import-export" models are covered by new framework
- [ ] Show "original row data" for each row when importing data
- [ ] Allow user to create new related model "in place"
- [ ] Add "context specific" importers for particular tables, rather than just in the admin center
- [ ] Improve efficiency of data import process (currently takes ~1 minute for 1,000 records)
- [ ] Impove user interface for data import - in particular, react hook efficiency
- [ ] Enable per-column specification of "default" values in the admin portal
- [ ] Documentation for new data import process

### CODE REVIEW

**Comment by matmair** [comment]
Very exciting! Is the removal of json/yaml a permanent change or something that was not in scope for the initial implementation?

**Comment by matmair** [comment]
@SchrodingersGat is this still slated for 1.0 or is 0.16 or sooner realistic?

**Comment by SchrodingersGat** [comment]
I'm hoping for 0.16.0 release. Now that the refactoring is done I have some time to commit to this. Soon it should be ready for a first round review.

**Comment by matmair** [comment]
Very cool. There is a lot of untested frontend code that can be removed once the new importer works.

**Comment by SchrodingersGat** [comment]
Yes once we can remove the old importer code, that will be fantastic.

**Comment by smithcv** [comment]
Can we merge this?

**Comment by matmair** [comment]
@smithcv this is not finalized and tested. Please allow the time needed to get it done properly with the resources the project has.

**Comment by SchrodingersGat** [comment]
@smithcv thanks for your interest but still a *lot* of work here. If you or your company would like to contribute to the development of this feature, you can provide [funding against this issue](https://github.com/inventree/InvenTree/issues/3101) :)

**Comment by matmair** [comment]
@SchrodingersGat I have created a branch that removes all import-export dependant things: https://github.com/matmair/InvenTree/tree/remove-imports

There are a lot of places where we use resource definitions for export - making this more generic (maybe as a view mixin?) would be great to remove code duplication

**Comment by SchrodingersGat** [comment]
### Data Export Profiling

The new *data export* functionality uses the optimized API serializers, and (at the current implementation) is already a huge improvement over the existing export tool which uses [django-import-export](https://django-import-export.readthedocs.io/en/latest/)

| API Endpoint | Record Count | Old Exporter | New Exporter | 
| --- | --- | --- | --- |
| `/api/part/` | 5,000 | *did not return after multiple minutes* | 17.3s , 11 DB hits |
| `/api/stock/` | 1,100 | 3.4s in 890 DB hits | 4.0s in 16 DB hits |
| `/api/stock/location/` | 20 | 0.4s in 38 DB hits |  0.5s in 4 DB hits |
| `/api/order/po/` | 10 | 0.4s in 48 DB hits | 0.5s in 23 DB hits |

However there is still some optimizing to be done here. Still need to profile where the time is being spent on the data export. Even though `django-import-export` uses a lot more DB hits their export is still faster in some instances.

**Comment by matmair** [comment]
@SchrodingersGat that looks very promising

File: src/frontend/src/components/importer/ImportDataSelector.tsx [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    return row?.errors[column.field] ?? [];
+  }, [row.errors, column.field]);
+
+  return (
+    <Group grow position="apart">
+      <Group grow style={{ flex: 1 }}>
+        <Stack spacing="xs">
+          <Text>{row.data[column.field]}</Text>
+          {cellErrors.map((error: string) => (
+            <Text size="xs" color="red">
```
Comment (wolflu05) [inline src/frontend/src/components/importer/ImportDataSelector.tsx:52]: Don't forget to add a key when rendering elements in a loop.

```suggestion
            <Text size="xs" color="red" key={error}>
```

**Comment by matmair** [comment]
@SchrodingersGat you need any assistance with this?

**Comment by SchrodingersGat** [comment]
I need to remember where I was up to and what the next steps might be. There is still a lot of work to go..

So far I have got a framework in place for creating an import session, and extracting data from table. This data is split into rows which get "serialized" against the matching serializer for validation and ingest.

### Major Tasks

- [ ] The user interface elements (in PUI) still need a lot of work
- [ ] Do we want to support this framework within the django admin interface?
- [ ] Still need to work out the final steps for validating data rows and then saving to the database
- [ ] No unit testing as yet

If you wanted to take a look at where I'm up to so far and provide input on the approach in general. I'm hoping to get back to this soon, now that the reporting has been refactored. It is a major blocker for a lot of outstanding issues.

File: src/frontend/src/hooks/UseImportSession.tsx [hunk 2]
```diff
@@ -9,6 +9,7 @@ import { useInstance } from './UseInstance';
  * Custom hook for managing the state of a data import session
  */
 
+// TODO: Load these values from the server?
 export enum ImportSessionStatus {
```
Comment (matmair) [inline src/frontend/src/hooks/UseImportSession.tsx:13]: @SchrodingersGat could this just be a generic state?

Comment (SchrodingersGat) [reply src/frontend/src/hooks/UseImportSession.tsx:13]: Defined via the API, you mean, like a "stock status" or similar?

Comment (matmair) [reply src/frontend/src/hooks/UseImportSession.tsx:13]: Yes, that would make it pretty flexible

**Comment by SchrodingersGat** [comment]
This is now progressed far enough to merge into master. Exporting data from tables (in CUI and PUI) now uses the new data export tool and is *much* faster. Data are serialized into the file using the existing DRF layer, which is very query efficient and reduces code duplication.

There is now also a "Data Import" section in the PUI admin page - which is functional, but still requires some work:

image


There are a lot of outstanding tasks / improvements to do (see outstanding tasks above) which will be pushed out into new issues.

The goal is to remove the reliance on `django-import-export` entirely.

**Comment by matmair** [comment]
@SchrodingersGat thank you for pushing this ahead! Are any of these goals blocking for 0.16.0 release? We should probably log and tag those separately to not lose track over the next month.

**Comment by SchrodingersGat** [comment]
I will make some separate issues soon, while it is still fresh on my mind.

**Pull request merged**